mypy added

to-sta opened the pull request:
<!---
Thank you for your pull request! 🚀
-->

### Contributor checklist

<!-- Please replace the empty checkboxes [] below with checked ones [x] accordingly. -->

- [x] This pull request is on a [separate branch](https://docs.github.com/en/get-started/quickstart/github-flow) and not the main branch

---

### Description

In this PR I have added mypy for static type checking including mypy plugins for django and drf.
The `requirements.txt` is updated to reflect the newly installed packages.

In order to test `mypy` i had to adjust `models.py` files. There were classes referenced as FK before they were declared and `ArrayField(IntegerField())` had missing arguments. 

Result of the test after the changes: 

```bash
events\models.py:31: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:44: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:63: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:71: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:79: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:87: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:95: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:103: error: Function is missing a type annotation  [no-untyped-def]
events\models.py:111: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:27: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:35: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:48: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:56: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:67: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:75: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:91: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:100: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:108: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:116: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:125: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:133: error: Function is missing a type annotation  [no-untyped-def]
entities\models.py:141: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:27: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:40: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:48: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:61: error: Function is missing a type annotation  [no-untyped-def]
content\models.py:69: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:19: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:29: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:50: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:58: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:66: error: Function is missing a type annotation  [no-untyped-def]
authentication\models.py:74: error: Function is missing a type annotation  [no-untyped-def]
entities\admin.py:6: error: Missing type parameters for generic type "ModelAdmin"  [type-arg]
authentication\admin.py:6: error: Missing type parameters for generic type "ModelAdmin"  [type-arg]
manage.py:7: error: Function is missing a return type annotation  [no-untyped-def]
manage.py:7: note: Use "-> None" if function does not return a value
manage.py:22: error: Call to untyped function "main" in typed context  [no-untyped-call]
events\serializers.py:6: error: Missing type parameters for generic type "ModelSerializer"  [type-arg]
entities\serializers.py:6: error: Missing type parameters for generic type "ModelSerializer"  [type-arg]
authentication\serializers.py:6: error: Missing type parameters for generic type "ModelSerializer"  [type-arg]
entities\views.py:14: error: Function is missing a type annotation  [no-untyped-def]
entities\views.py:26: error: Function is missing a type annotation  [no-untyped-def]
entities\views.py:35: error: Function is missing a type annotation  [no-untyped-def]
entities\views.py:45: error: Function is missing a type annotation  [no-untyped-def]
Found 44 errors in 11 files (checked 42 source files)
```

<ins>Additionally:</ins>
I added `venv/` to the .gitignore file to prevent commiting my virtual environment.

List of packages installed:
- [mypy](https://github.com/python/mypy)
- [django-stub](https://github.com/typeddjango/django-stubs)
- [djangorestframework-stubs](https://github.com/typeddjango/djangorestframework-stubs)

<!--
Describe briefly what your pull request proposes to change. Especially if you have more than one commit, it is helpful to give a summary of what your contribution is trying to solve.

Also, please describe shortly how you tested that your change actually works.
-->

### Related issue

<!--- activist prefers that pull requests be related to already open issues. -->
<!--- If applicable, please link to the issue by replacing ISSUE_NUMBER with the appropriate number below. -->
<!--- Feel free to delete this section if this does not apply. -->

- #329

andrewtavis commented on backend/mypy.ini:
```diff
@@ -0,0 +1,9 @@
+strict = True
+
+[mypy.plugins.django-stubs]
+django_settings_module = backend.settings
```
We need a blank line at the end of the file :)

to-sta replied:
Same here :). Btw i think i can implement pre-commit if you are intrested. I started using it and it has a end-of-line fixer hook that would take care of this and it also can run black to get rid of the issue  above before submitting a PR

andrewtavis replied:
For end of line formatting I think that would make sense, @to-sta! Thanks for the offer to implement!

andrewtavis reviewed (commented):
Another comment on this, @to-sta, looks like the `.editorconfig` file was removed. Is this because the other files added in recent PRs for Prettier and other workflows are replacing it?

to-sta commented on backend/authentication/models.py:
```diff
@@ -13,9 +13,9 @@
 
 class Support(models.Model):
-    supporter_type = models.ForeignKey(SupportEntityType, on_delete=models.CASCADE)
+    supporter_type = models.ForeignKey('SupportEntityType', on_delete=models.CASCADE)
```
Damn i missed that, I will format that again.

to-sta commented:
Added `.editorconfig` back in again, my bad that was by mistake. Also added `*.mypy_cache` to the `.gitignore` and `backend/.dockerignore` to avoid commiting/containerizing it in the future.

andrewtavis reviewed (approved):
Thanks for the quick changes, @to-sta! I' check in in the issue to see if this is good to close or if we need to wait on it to see how all's working 😊
